### PR TITLE
flux-ping refactor and cleanup

### PR DESCRIPTION
**Pull request by chu11**
Per discussion in PR #860 and some offline discussion, cleanup flux-ping to use jansson API.  I'm not 100% done as I noticed a few things, but thought I'd start w/ this PR.

Lots of mini-cleanup.  Notable changes:

For consistency to ping(8), change ```--delay``` to ```--interval``` in flux-ping.  This is an outright breakage of old command line but thought important to make it consistent and now's the time to do it.  acc62ef7337fbdbd7ed1dca527e0c99a3a21fcab

Refactor ```flux_rpc_multi``` and add a ```flux_rpcf_multi``` jansson equivalent.  Stuck typdef of callback in the middle of the file, right before usage.  It's not the style I normally use, but noticed it elsewhere.  Could cleanup usage all around if not liked.  96171838a9a299a8e6cd2044796bd8b418134830 & 7cab3c6c8679f6ba935aa7ccebb4f69183dd9357

Use jansson API in ```flux-ping```.  Not as cleaned up as I hoped. It's a complete wash when it comes to LOC.  Perhaps need to just call ```flux_rpcf_multi``` regardless if it's a single or multi-call.  Haven't quite figured out the right logic for that yet.  Perhaps a rank of ```"any"``` could be another key that can be accepted by the ```flux_rpc_multi``` functions? 87ee115f6b1a3ad475e3c2a4984e851ae105d38f

I continue to use ```int64_t``` where it would be preferable to use ```json_int_t```, but there's some header collisions between libjson-c and jansson.  Can't remove ```shortjson.h``` header, because it's needed for ```tstat.h```.  So should refactor that somehow.

Another thing I noticed is that in ```message.c``` we got this check for ```json_int_t```.  Now that ```json_int_t``` can be used multiple places, I'm thinking we should stick it in ```configure``` and just fail if longs aren't 8 bytes.

```
/* Ensure %I format for jansson pack/unpack works with int64_t
 */
#if JSON_INTEGER_IS_LONG_LONG
# if SIZEOF_LONG_LONG != 8
# error json_int_t is not 64 bits
# endif
#else
# if SIZEOF_LONG != 8
# error json_int_t is not 64 bits
# endif
#endif
```

### CODE REVIEW

**Comment by garlick** [comment]
Couple quick comments:

> Can't remove shortjson.h header, because it's needed for tstat.h. So should refactor that somehow.

My first thought would be to drop `tstat_json()` from tstat.[ch].  Since the `tstat_t` struct is exposed, the few places that need a JSON encoder can just do it themselves using whatever is convenient in that context.

> Another thing I noticed is that in message.c we got this check for json_int_t. Now that json_int_t can be used multiple places, I'm thinking we should stick it in configure and just fail if longs aren't 8 bytes.

That sounds fine with me.  The idea was that `int64_t` would be used in place of `json_int_t` in Flux API calls (and I think the man pages document it that way).

**Comment by chu11** [comment]
> That sounds fine with me. The idea was that int64_t would be used in place of json_int_t in Flux API calls (and I think the man pages document it that way).

Ahh, had missed that it had been documented with ```int64_t```.  Reminds me I need to add the manpages for ```flux_rpcf_multi``` too.

**Comment by garlick** [comment]
The `flux_rpcf_multi()` control flow is a bit hard to follow (internally in rpc.c) with the introduction of the callbacks.  I'm wondering if the refactoring was worth it or if it would be clearer to replicate the shared code, or abstract some of it to a single shared function?  I haven't worked through it in detail so maybe where you ended up is the best way.  I find it confusing though.  What's your opinion?

By the way IMHO it is fine to put callback typedefs at the top.  I don't personally have a strong opinion on that one, or maybe would lean towards doing it at the top also.

**Comment by chu11** [comment]
The refactor and ```flux_rpcf_multi()``` sort of came about due to not wanting to duplicate large amounts of the code and wanting to handle the cleanup of ```va_list``` neatly.  Here's a pseudo-code of the logic when ```va_list``` was added into the middle of the original ```flux_rpc_multi()```:

```
flux_rpcf_multi(const char *fmt, ...)
{
    // bunch of setup including setting up w/ va_start()
    va_start (ap, fmt);
    for (i = 0; i < count; i++) {
        va_copy (ap_cpy, ap);
        if (rpc_send (fmt, ap_cpy) < 0) {
            va_end (ap_cpy);
            goto error;
         }
         va_end (ap_cpy);
    }
    // various cleanup, including gotta call va_end()
    va_end (ap);
    return rpc;
error:
    // cleanup and destroy everything, including va_end()
    va_end (ap);
    return NULL;
}
```

The ```va_*``` calls are sort of all over the place.  The callbacks were to try and isolate all the ```va_*``` calls into a smaller snippet.  Fully acknowledged maybe it's not worth it.  We're certainly not at a "rule of three" situation, so perhaps an outright duplicate of the shared code would still be ok.  I could really go either way, but if you've already find it difficult to grok, then perhaps code duplication would be better.

**Comment by garlick** [comment]
Is there a way to avoid the repeated creation of the json payload and the varargs copying by creating the message once and then in the loop, calling `rpc_request_prepare()` to set the nodeid/matchtag, and `flux_send()`, then destroy the message outside the loop?

I wonder if that would refactor a little easier?

**Comment by chu11** [comment]
Ahh, that's a good idea.  Actually, at worst I can just build the json-str myself and just pass it directly into ```flux_rpc_multi```.  That'd be way better.

**Comment by chu11** [comment]
After four spinoff PRs, (#902, #904, #909, #911) here is final flux-ping cleanup PR.

Discounting liboptparse cleanup, the LOC are largely unchanged, but I think the code is easier to follow w/ the jansson changes and the ```rank_count``` variable (and I believe it fixes some corner cases as well).

**Comment by garlick** [comment]
Nice!  Thanks very much for working through this cleanup, and especially for the test cases and well factored PR's, which I know are a pain to work on but in the long run are well worth it IMHO.
